File: DS_and_Algos/Javascript/search_in_2D_matrix.js
var searchMatrix = function (matrix, target) {
  if (matrix[0] === undefined) {
    return false;
  }

  let min = 0;
  let max = matrix[0].length - 1;

  while (min <= max) {
    let mid = Math.floor((min + max) / 2);

    if (matrix[0][mid] === target) {
      return true;
    } else if (matrix[0][mid] > target) {
      max = mid - 1;
    } else min = mid + 1;
  }

  matrix.splice(0, 1);
  return searchMatrix(matrix, target);
};

console.log(
  "searchMatrix",
  searchMatrix(
    [
      [1, 3, 5, 7],
      [10, 11, 16, 20],
      [23, 30, 34, 50],
    ],
    33
  )
);

File: DS_and_Algos/Python/Algorithms/BinarySearch/search_in_2D_matrix.py
class Solution:
    def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
        cols, rows, m = len(matrix), len(matrix[0]), len(matrix[0]) - 1
        left, right = 0, cols - 1

        # search where the number belongs
        while left < right:

            mid = (left + right) // 2

            if matrix[mid][m] < target:
                left = mid + 1
            
            else:
                right = mid

        targetIndex = right
        left, right = 0, rows - 1

        # search through the founded matrix
        while left <= right:

            mid = (left + right) // 2

            if matrix[targetIndex][mid] == target:
                return True

            if matrix[targetIndex][mid] < target:
                left = mid + 1
            else:
                right = mid - 1

        return False
        
ExampleSolution = Solution()
print(ExampleSolution.searchMatrix(
   [[1],[3]], 1
    )
)